### DISCUSSION TITLE
RemoteControl / HotReload with Skia.Gtk, Embedded Linux and VS

### DISCUSSION THREAD

**Discussion by brinko99** [unoplatform/uno#14308] [Q&A]
We're developing for Skia.Gtk on embedded Linux. Is it possible to get RemoteControl/HotReload working on Linux embedded? We develop with Visual Studio and can debug over SSH.

I note that documentation mentions HotReload support under Linux with VisualStudio Code, but I presume this applies when running VS Code on the Linux machine itself.

On startup, Uno reports "Failed to get remote control server port from the IDE for endpoint..." with various IP addresses.

Thanks

**Comment by jeromelaban** [comment]
Thanks for the question! If you're using a setup with Remote SSH with VS Code, the tooling should be working properly, where the "DevServer" tooling is running on the machine where to SSH-ing to.

In VS Code, can you show what the output of "Uno Platform - Hot Reload"? You can also enable tracing in the running app by setting the [minimum log level to `Trace`](https://github.com/unoplatform/uno.templates/blob/a44dff150cef982b7681f81786540db90e667c14/src/Uno.Templates/content/unoapp/MyExtensionsApp._1.Base/AppHead.xaml.cs#L72) and the [DevServer client to Trace](https://github.com/unoplatform/uno.templates/blob/a44dff150cef982b7681f81786540db90e667c14/src/Uno.Templates/content/unoapp/MyExtensionsApp._1.Base/AppHead.xaml.cs#L101) as well.

Also, you can find [additional troubleshooting info here](https://aka.platform.uno/hot-reload).

**Reply by brinko99** [reply]
We are using Visual Studio, installed on Windows, not VS Code. Is RemoteControl applicable to Visual Studio? I don't believe there's any possibility of supporting Remote SSH with VS Code using the VS Code Server on our target.

"DevServer" is not a term I'm familiar with. It looks like this has replaced RemoteControl, correct? I should mention that we're on .NET 6 and Uno 4.x currently and too near ship to move to Uno 5 in the short term.

Thanks as always

**Reply by jeromelaban** [reply]
Got it, thanks! Yes, DevServer is the new name for RemoteControl in Uno 5. So if you're inside VS, this means that your device needs to have network access to your Windows machine and possibly through the firewall. You'll also find the same `Uno Platform` output window that details what is happening to the server component of hot reload, to find out if there are any connections happening.

If none happens, which is likely with the message you've posted from the running app, it may only be a network connectivity issue. Try to temporarily disable your firewall to see if it helps. If it does, you can add an exclusion for the RemoteControl binary.

**Reply by brinko99** [reply]
Thanks for the clarification...

From the Uno Platform output window I see the below.  The project does indeed contain `Uno.WinUI.RemoteControl` and it has been restored, built, and runs.
```
16:02:39:352	[INFO] The project does not contain the Uno.WinUI.DevServer nor Uno.UI.DevServer package.
16:02:54:591	[INFO] The Uno.WinUI.RemoteControl or Uno.UI.RemoteControl package has not been restored yet, retrying...
16:07:51:548	[ERROR] Failed to retrieve the solution's loaded NuGet packages. Make sure a full NuGet restore has completed properly and reload the solution.
16:07:56:586	[ERROR] Failed to retrieve the solution's loaded NuGet packages. Make sure a full NuGet restore has completed properly and reload the solution.
....
```

I'm not able to temporarily disable the firewall on this machine; we're locked down and I'll need to request an exception for the port.

On my personal machine, I have more success and can get DevServer running. However, even though I see that a change was recognized ("...MainPage.xaml changed"), I see no effect in the running application. See the attached Uno Platform output. This is tested purely on Windows with Skia.Gtk with Visual Studio and the app running locally. When I attach the debugger to the running app,  HotReload will not accept simple XAML changes.

This leaves me with the following questions:

1) What explains the warnings and errors above regarding NuGet packages?
2) What are the specific firewall rules that I need to install? It appears that. It appears that it is Uno.UI.RemoteControl.Host.dll on port 54823 that needs to be excepted. Does this port change and can it be customized?
3) When using Visual Studio with DevServer. Should the debugger be attached? Is this expected to work with the debugger attached remotely as well when I'm testing on Linux?

Sorry for the deluge of questions; thanks again!

[UnoRemoteControlLog.txt](https://github.com/unoplatform/uno/files/13290590/UnoRemoteControlLog.txt)

**Reply by jeromelaban** [reply]
Thanks for the updates and no worries about the questions.

> `16:07:56:586	[ERROR] Failed to retrieve the solution's loaded NuGet packages. Make sure a full NuGet restore has completed properly and reload the solution.`

This is very strange. Does it give the same error when reloading Visual Studio?

> I'm not able to temporarily disable the firewall on this machine; we're locked down and I'll need to request an exception for the port.
> What are the specific firewall rules that I need to install? It appears that. It appears that it is Uno.UI.RemoteControl.Host.dll on port 54823 that needs to be excepted. Does this port change and can it be customized?

From Visual Studio, there's no direct way for providing a specific port for the RC/DevServer from the UI though you should be able to set it by adding this in your app head:
```xml
<PropertyGroup>
   <UnoRemoteControlPort>56789</UnoRemoteControlPort>
</PropertyGroup>
```

> When using Visual Studio with DevServer. Should the debugger be attached? Is this expected to work with the debugger attached remotely as well when I'm testing on Linux?

In Uno 4.x, Hot Reload is XAML based-only (XamlReader) and has less features that the one in Uno 5 (C# HotReload based which includes XAML as well). Both can run with and without debugger when running on the local machine, though I don't know in your context with remote debugging with SSH if without debugger it will work for C#-based hot reload. 

The question is: Are you able to do C# HotReload with remote debugging? If you can, then Uno 5's hot reload is likely to be better suited for what you need, as the DevServer is not needed to reload (even if there are connection errors).

**Reply by brinko99** [reply]
I believe I've determined the root cause of the issue that produced:
`The project does not contain the Uno.WinUI.DevServer nor Uno.UI.DevServer package`

I keep my NugetCache on a secondary drive, configured with `NuGet.config` (below). Uno does not seem to support this; I believe it's hardcoded to look in `%AppData%`.

`<add key="globalPackagesFolder" value="D:\NugetCache" />`

Now that the RemoteControl.Host is running properly, HotReload of XAML changes to a GTK project does work locally. I'm testing under Uno 5, .NET 7, with the latest extension.

Next, I'm trying to get it to work on an app running on my embedded Linux target. I am able to see that a connection is made between the app and RemoteControl.Host. However, though changes are recognized by the host, the running app is not updated. I've tried the following:

* Debugger Not Attached
  * Make XAML change and Save (no effect) (log attached)
* Debugger Attached
  * Make XAML change and Save (no effect)
  * Changes to C# at breakpoint (no effect)
  * VS does not show HotReload enabled

My remaining question is, what aspects of RemoteControl / HotReload are expected to work in this situation? It sounds like the save-based, no debugger-attached scenario should work.

Thanks,

**Reply by jeromelaban** [reply]
Thanks for the update.

> I keep my NugetCache on a secondary drive, configured with NuGet.config (below). Uno does not seem to support this; I believe it's hardcoded to look in %AppData%.

This is quite interesting, because there's no hard coding of paths in this section, but we'll need to investigate why this may happen.

> VS does not show HotReload enabled

That's very interesting as well. This would mean that the remote VS debugger does not support hot reload in the SSH context. I wonder, how do you start the app on your device? Do you start it manually, then attach to the running app?

If you're attaching to the app (and it's completely undocumented and untested), let's try something.

Add the following property to the skia.gtk project:
```xml
<PropertyGroup>
   <UnoHotReloadMode>MetadataUpdates</UnoHotReloadMode>
</PropertyGroup>
```

Then try running the app this way:
```bash
export DOTNET_MODIFIABLE_ASSEMBLIES=debug
dotnet yourapp.dll
```

This will try to bypass Visual Studio's hot reload and use Uno's.

Try collecting Hot reload logs in Uno and VS, and post them here.

Thanks!

**Reply by brinko99** [reply]
I've done some more testing per your suggestions. Yes, I am starting the app manually and attaching the VS debugger through SSH (I don't know of another way to launch the app remotely).

I've added the property `UnoHotReloadMode` and set the variable `DOTNET_MODIFIABLE_ASSEMBLIES`. I don't believe I'm seeing any difference. Here are the application logs and Uno extension logs (each file contains both).

[UnoRemoteControlLog (No Debugger).txt](https://github.com/unoplatform/uno/files/13322832/UnoRemoteControlLog.No.Debugger.txt)
[UnoRemoteControlLog (With Debugger).txt](https://github.com/unoplatform/uno/files/13322833/UnoRemoteControlLog.With.Debugger.txt)

When I make a XAML change, I see the "Presentation changed" message from the host, and the client shows "Skipping file reload".

Other notes:
* I'm doing a self-contained debug publish and launching the app directly, not through `dotnet` (we have NET6 on the box only, this test is being done with NET7)
* Never does the VS Hot Reload icon become available. The Hot Reload window remains empty
* I've tried setting the variables `DisableXbfLineInfo=false` and `XamlDebuggingInformation=true` per MS' troubleshooting tips not knowing if this applies to WinUI3 / Uno
* Looking at the source for `ClientHotReloadProcessor` it appears I should see a trace message that says "Forced Hot Reload Mode" since `UnoHotReloadMode` was set. I do not.
* When I try "Edit and Continue" at a breakpoint, I see the "Presentation changed" and "Skipping file reload" messages as well, with no effect.

I'll be out for the next couple of weeks and look forward to picking this up then.

Thanks

**Reply by brinko99** [reply]
Following up on this. @jeromelaban; you'd asked me to perform some additional experiments here and capture logs. Do you have any thoughts on the results?

Thanks

**Reply by jeromelaban** [reply]
No specific progress has been done on this, though the only search path here is that hot reload can only work when the debugger is not attached. Once the app is started and connected back to the DevServer, hot reload should be working.
